src/main/java/com/algorand/algosdk/util/abi/TypeArrayDynamic.java

```diff
@@ -0,0 +1,57 @@
+package com.algorand.algosdk.util.abi;
+
+import com.algorand.algosdk.util.Encoder;
+
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+
+public class TypeArrayDynamic extends Type {
+    public final Type elemType;
+
+    public TypeArrayDynamic(Type elemType) {
+        this.elemType = elemType;
+    }
+
+    @Override
+    public String toString() {
+        return this.elemType.toString() + "[]";
+    }
+
+    @Override
+    public boolean isDynamic() {
+        return true;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof TypeArrayDynamic)) return false;
+        return this.elemType.equals(((TypeArrayDynamic) o).elemType);
+    }
+
+    @Override
+    public byte[] encode(Object o) {
+        if (!(o instanceof Object[]))
+            throw new IllegalArgumentException("cannot infer type for abi dynamic array encode");
+        Object[] objArray = (Object[]) o;
+
+        byte[] castedEncode = Type.castToTupleType(objArray.length, this.elemType).encode(objArray);
+        byte[] lengthEncode = Encoder.encodeUintToBytes(BigInteger.valueOf(objArray.length), 2);
```

Comment:
Did we want to use a constant (ABI_LENGTH_BYTE_SIZE or something) instead of 2 for the length bytes?